sycl/include/sycl/property_list.hpp

```diff
@@ -44,13 +44,12 @@ class property_list : protected detail::PropertyListBase {
     ctorHelper(Props...);
   }
 
-  template <typename PropT> PropT get_property() const {
-    if (!has_property<PropT>())
-      throw sycl::invalid_object_error("The property is not found",
-                                       PI_ERROR_INVALID_VALUE);
-
-    return get_property_helper<PropT>();
-  }
+  template <typename PropT> PropT get_property() const;// {
+  //   if (!has_property<PropT>())
+  //     throw sycl::invalid_object_error("The property is not found",
+  //                                      PI_ERROR_INVALID_VALUE);
+  //   return get_property_helper<PropT>();
+  // }
 
```

Comment:
@steffenlarsen 
I created property_list.cpp and moved this code to property_list.cpp.
This caused lots of undefined symbol get_property().
AFAIK, we can't move a templated method to the cpp file because of this issue.
If you can come up with an idea how to solve this issue, I can try it.
Below is just a few examples.
```
/usr/bin/ld: tools/sycl/source/CMakeFiles/sycl_object.dir/stream.cpp.o: in function `sycl::_V1::ext::oneapi::property::queue::priority_low syc
l::_V1::stream::get_property<sycl::_V1::ext::oneapi::property::queue::priority_low>() const':
stream.cpp:(.text._ZNK4sycl3_V16stream12get_propertyINS0_3ext6oneapi8property5queue12priority_lowEEET_v+0x10): undefined reference to `sycl::_
V1::ext::oneapi::property::queue::priority_low sycl::_V1::property_list::get_property<sycl::_V1::ext::oneapi::property::queue::priority_low>()
 const'
/usr/bin/ld: tools/sycl/source/CMakeFiles/sycl_object.dir/stream.cpp.o: in function `sycl::_V1::ext::oneapi::property::queue::priority_high sy
cl::_V1::stream::get_property<sycl::_V1::ext::oneapi::property::queue::priority_high>() const':
stream.cpp:(.text._ZNK4sycl3_V16stream12get_propertyINS0_3ext6oneapi8property5queue13priority_highEEET_v+0x10): undefined reference to `sycl::
_V1::ext::oneapi::property::queue::priority_high sycl::_V1::property_list::get_property<sycl::_V1::ext::oneapi::property::queue::priority_high
>() const'
/usr/bin/ld: tools/sycl/source/CMakeFiles/sycl_object.dir/stream.cpp.o: in function `sycl::_V1::ext::oneapi::property::queue::priority_normal
sycl::_V1::stream::get_property<sycl::_V1::ext::oneapi::property::queue::priority_normal>() const':
stream.cpp:(.text._ZNK4sycl3_V16stream12get_propertyINS0_3ext6oneapi8property5queue15priority_normalEEET_v+0x10): undefined reference to `sycl
::_V1::ext::oneapi::property::queue::priority_normal sycl::_V1::property_list::get_property<sycl::_V1::ext::oneapi::property::queue::priority_
normal>() const'
/usr/bin/ld: tools/sycl/source/CMakeFiles/sycl_object.dir/stream.cpp.o: in function `sycl::_V1::property::buffer::mem_channel sycl::_V1::strea
m::get_property<sycl::_V1::property::buffer::mem_channel>() const':
stream.cpp:(.text._ZNK4sycl3_V16stream12get_propertyINS0_8property6buffer11mem_channelEEET_v+0x25): undefined reference to `sycl::_V1::propert
y::buffer::mem_channel sycl::_V1::property_list::get_property<sycl::_V1::property::buffer::mem_channel>() const'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.

```

Reply:
That is right. If they are templated, we can't do that. What I suggest is adding a new function that goes in place of `     throw sycl::invalid_object_error("The property is not found", PI_ERROR_INVALID_VALUE);` and which throws from inside the runtime. It could be as simple as:

```
// .hpp
void ThrowSYCL:ExceptionHelper(errc, const char *);

// .cpp
void ThrowSYCL:ExceptionHelper(errc ErrorCode, const char *ErrorMessage) {
  throw sycl::exception(ErrorCode, ErrorMessage);
}
```
And then some comments saying it is to avoid cyclic references. It's not perfect as the compiler can't know that it always throws, but in this case it is fine.

Reply:
Thanks.